File: cpp/src/Card.cpp
#include "Card.hpp"
using namespace std;

Card::Card(Rank r, Suit s) : rank(r), suit(s) {}

Rank Card::getRank() const {
    return rank;
}

Suit Card::getSuit() const {
    return suit;
}

bool Card::operator<(const Card& other) const {
    if (rank != other.rank) {
        return static_cast<int>(rank) < static_cast<int>(other.rank);
    }

    return static_cast<int>(suit) < static_cast<int>(other.suit);
}

bool Card::operator>(const Card& other) const {
    return other < *this;
}

bool Card::compareByRank(const Card& a, const Card& b) {
    return a.rank < b.rank;
}

bool Card::compareBySuit(const Card& a, const Card& b) {
    if (a.suit != b.suit) {
        return a.suit < b.suit;
    }

    return a.rank < b.rank;
}

string Card::toString() const {
    static const string rankStrings[] = {
        "2", "3", "4", "5", "6", "7", "8", "9", "10",
        "J", "Q", "K", "A"
    };

    static const string suitStrings[] = {
        "♠", "♥", "♣", "♦"
    };

    return rankStrings[static_cast<int>(rank)] + suitStrings[static_cast<int>(suit)];
}

File: java/src/poker/Card.java
package com.poker;

public class Card implements Comparable<Card> {
    private final Rank rank;
    private final Suit suit;
    
    public Card(Rank rank, Suit suit) {
        if (rank == null || suit == null) {
            throw new IllegalArgumentException("Rank and suit cannot be null");
        }
        this.rank = rank;
        this.suit = suit;
    }
    
    public Rank getRank() {
        return rank;
    }
    
    public Suit getSuit() {
        return suit;
    }
    @Override
    public int compareTo(Card other) {
        return Integer.compare(this.rank.getValue(), other.rank.getValue());
    }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Card)) return false;
        Card other = (Card) obj;
        return this.rank == other.rank && this.suit == other.suit;
    }
    
    @Override
    public int hashCode() {
        return 31 * rank.hashCode() + suit.hashCode();
    }
    
    @Override
    public String toString() {
        return rank.getSymbol() + suit.getSymbol();
    }
    
    public String toDisplayString() {
        return rank.getSymbol() + suit.getUnicodeSymbol();
    }
    
    public static Card fromString(String cardStr) {
        if (cardStr == null || cardStr.length() < 2) {
            throw new IllegalArgumentException("Invalid card string: " + cardStr);
        }
        
        String rankStr = cardStr.substring(0, cardStr.length() - 1);
        String suitStr = cardStr.substring(cardStr.length() - 1);
        
        Rank rank = Rank.fromSymbol(rankStr);
        Suit suit = Suit.fromSymbol(suitStr);
        
        return new Card(rank, suit);
    }
}


